src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs

```diff
@@ -111,5 +126,115 @@ protected override void ExecuteCore()
             _filesWritten.Add(new TaskItem(DepsFilePath));
 
         }
+
+        private void LoadFilesToSkip()
+        {
+            foreach (var fileToSkip in FilesToSkip)
+            {
+                string packageId, packageSubPath;
+                ConflictResolution.NuGetUtilities.GetPackageParts(fileToSkip.ItemSpec, out packageId, out packageSubPath);
+
+                if (String.IsNullOrEmpty(packageId) || String.IsNullOrEmpty(packageSubPath))
+                {
+                    continue;
+                }
+
+                var itemType = fileToSkip.GetMetadata(nameof(ConflictResolution.ConflictItemType));
+                var packagesWithFilesToSkip = (itemType == nameof(ConflictResolution.ConflictItemType.Reference)) ? compileFilesToSkip : runtimeFilesToSkip;
+
+                HashSet<string> filesToSkipForPackage;
+                if (!packagesWithFilesToSkip.TryGetValue(packageId, out filesToSkipForPackage))
+                {
+                    packagesWithFilesToSkip[packageId] = filesToSkipForPackage = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+                }
+
+                filesToSkipForPackage.Add(packageSubPath);
+            }
+        }
+
+        private DependencyContext TrimFilesToSkip(DependencyContext sourceDeps)
+        {
+            return new DependencyContext(sourceDeps.Target,
+                                         sourceDeps.CompilationOptions,
+                                         TrimCompilationLibraries(sourceDeps.CompileLibraries),
+                                         TrimRuntimeLibraries(sourceDeps.RuntimeLibraries),
+                                         sourceDeps.RuntimeGraph);
+        }
+
+        private IEnumerable<RuntimeLibrary> TrimRuntimeLibraries(IReadOnlyList<RuntimeLibrary> runtimeLibraries)
+        {
+            foreach (var runtimeLibrary in runtimeLibraries)
+            {
+                HashSet<string> filesToSkip;
+                if (runtimeFilesToSkip.TryGetValue(runtimeLibrary.Name, out filesToSkip))
+                {
+                    yield return new RuntimeLibrary(runtimeLibrary.Type,
```

Comment:
@wli3 - you are right, you can't remove a library based solely on "no assets".  You'd have to take into account its dependencies as well.

> Could you give me pointer to the code to find out?

A pointer to what code?  There are a few consumers of the deps.json file, most notably is the MVC code:

https://github.com/aspnet/Mvc/blob/c8a1bb69142792d1046044bde7082b4e3ac8ac27/src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultAssemblyPartDiscoveryProvider.cs

As you can see, MVC will walk all the dependencies looking for assemblies that eventually depend on an MVC assembly.  So yes, if a library has a dependency to another library that eventually has an asset, all the libraries in the dependency chain shouldn't be trimmed.